File: examples/example_stdin_stdout_program.js
const arg1 = process.argv[2];
const arg2 = process.argv[3];

const uppercase = arg1 === "--uppercase";
const intentionalTimeout = arg2 === "--intentional-timeout";
const intentionalCrash = arg2 === "--intentional-crash";

console.log({ uppercase, intentionalTimeout, intentionalCrash });

const wait = (duration = 1000) => new Promise((res, _rej) => setTimeout(res, duration));

async function sendMessage(total) {
	if (total.trimEnd().endsWith("2")) {
		if (intentionalTimeout) await wait(3000);
		if (intentionalCrash) throw Error("CRASH!!!");
	}

	const bothChannels = total.includes("stderr");

	for (const line of total.split("\n")) {
		const chunk = uppercase ? line.toUpperCase() : line;
		const stream = line.endsWith("on stderr") ? Bun.stderr : Bun.stdout;

		const writer = stream.writer();
		writer.write(chunk);
		writer.write("\n");
		writer.flush();

		if (bothChannels) await wait(50);
	}
}

console.log("start");
let buffer = "";
for await (const line of console) {
	if (line == "close") break;

	if (line == "end") {
		await sendMessage(buffer);
		console.log("end");
		buffer = "";
		continue
	}

	buffer += line;
	buffer += "\n";
}

console.log("finished");

File: examples/example_stdin_stdout_program.rs
use std::env;
use std::io::{self, BufRead};

fn main() {
    let is_uppercase = env::args().any(|flag| flag.as_str() == "--uppercase");
    let intentional_crash = env::args().any(|flag| flag.as_str() == "--intentional-crash");
    let stdin = io::stdin();
    let mut buf = Vec::new();

    println!("start");

    for line in stdin.lock().lines() {
        let Ok(line) = line else { break };

        if line == "close" {
            if !buf.is_empty() {
                eprintln!("no end to message {buf:?}");
            }
            break;
        }

        if line == "end" {
            let output = String::from_utf8_lossy(&buf);
            for line in output.lines() {
                if intentional_crash && line.trim_end().ends_with("2") {
                    panic!("CRASH!!!");
                }
                let output = if is_uppercase {
                    std::borrow::Cow::Owned(line.to_uppercase())
                } else {
                    std::borrow::Cow::Borrowed(line)
                };
                if line.trim_end().ends_with("on stderr") {
                    std::thread::sleep(std::time::Duration::from_millis(50));
                    eprintln!("{output}");
                    std::thread::sleep(std::time::Duration::from_millis(50));
                } else {
                    println!("{output}");
                }
            }
            println!("end");
            buf.clear();
            continue;
        }

        buf.extend_from_slice(line.as_bytes());
        buf.push(b'\n');
    }

    // println!("Finished!");
}
